File: linked_list_operations/ruby/linked_list.rb
class Node
  attr_accessor :value, :next_node

  def initialize(value, next_node = nil)
    @value = value
    @next_node = next_node
  end
end

class LinkedList
  attr_accessor :head

  def initialize
    @head = nil
  end

  # Add a new node at the end of the list
  def append(value)
    new_node = Node.new(value)
    if @head.nil?
      @head = new_node
    else
      current = @head
      current = current.next_node while current.next_node
      current.next_node = new_node
    end
  end

  # Add a new node at the beginning of the list
  def prepend(value)
    @head = Node.new(value, @head)
  end

  # Remove a node by value
  def delete(value)
    return @head = @head.next_node if @head.value == value

    current = @head
    while current.next_node && current.next_node.value != value
      current = current.next_node
    end

    current.next_node = current.next_node.next_node if current.next_node
  end

  # Search for a node by value
  def search(value)
    current = @head
    while current
      return true if current.value == value
      current = current.next_node
    end
    false
  end

  # Traverse the list and print each value
  def print_list
    current = @head
    while current
      print "#{current.value} -> "
      current = current.next_node
    end
    puts "nil"
  end
end

# Example usage:
list = LinkedList.new
list.append(10)
list.append(20)
list.append(30)
list.prepend(5)
list.print_list
puts "Does 20 exist in the list? #{list.search(20)}"
list.delete(20)
list.print_list

File: linked_list_operations/java/LinkedList.java
class LinkedList {

    // Node class representing each element in the linked list
    static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    Node head; // Head of the list

    // Method to insert a new node at the end of the list
    public void insert(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
        } else {
            Node last = head;
            while (last.next != null) {
                last = last.next;
            }
            last.next = newNode;
        }
    }

    // Method to delete a node with a specific value
    public void delete(int data) {
        Node current = head;
        Node previous = null;

        // If the head node itself holds the key to be deleted
        if (current != null && current.data == data) {
            head = current.next; // Change head
            return;
        }

        // Search for the key to be deleted, keep track of the previous node
        while (current != null && current.data != data) {
            previous = current;
            current = current.next;
        }

        // If the key was not present in the linked list
        if (current == null) {
            return;
        }

        // Unlink the node from the linked list
        previous.next = current.next;
    }

    // Method to display the contents of the linked list
    public void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        // Inserting elements
        list.insert(10);
        list.insert(20);
        list.insert(30);
        list.insert(40);

        // Displaying the list
        System.out.print("Linked List: ");
        list.display();

        // Deleting an element
        list.delete(20);
        System.out.print("After deleting 20: ");
        list.display();
    }
}
